third_party/move/move-vm/types/src/values/function_values_impl.rs

```diff
@@ -166,10 +167,13 @@ impl<'d, 'c> serde::de::Visitor<'d> for ClosureVisitor<'c> {
         let ty_args = read_required_value::<_, Vec<TypeTag>>(&mut seq)?;
         let mask = read_required_value::<_, ClosureMask>(&mut seq)?;
 
-        let num_captured_values = mask.captured_count() as usize;
-        let mut captured_layouts = Vec::with_capacity(num_captured_values);
-        let mut captured = Vec::with_capacity(num_captured_values);
-        for _ in 0..num_captured_values {
+        // Note: we explicitly DO NOT allocate capacity because deserialization is recursive. If
+        // the first captured argument is a function value, we will end up allocating the capacity
+        // one more time, and so on... which may lead to out-of-memory bugs.
```

Comment:
Really? Wouldn't that oom be captured by our depth check? With 128 * MAX_CAPTURE it shouldn't OOM?

Otherwise we are optimizing here for the pathological attack case. The 99.9% case would not have recursive closures.